libraries/utils/src/render-lines-on-sp.ts

```diff
@@ -34,15 +39,17 @@ export class ScatterPlotLinesRenderer {
         this.sp.render(this.canvas.getContext('2d') as CanvasRenderingContext2D);
     }
 
-    constructor(sp: DG.ScatterPlotViewer, xAxis: string, yAxis: string, lines: ILines) {
+    constructor(sp: DG.ScatterPlotViewer, xAxis: string, yAxis: string, lines: ILineSeries) {
         this.sp = sp;
-        this.xAxisValues = this.sp.dataFrame!.columns.byName(xAxis).toList();
-        this.yAxisValues = this.sp.dataFrame!.columns.byName(yAxis).toList();;
+        this.xAxisCol = this.sp.dataFrame!.columns.byName(xAxis);
+        this.yAxisCol = this.sp.dataFrame!.columns.byName(yAxis);
         this.lines = lines;
         this.canvas = this.sp.getInfo()['canvas'];
         this.ctx = this.canvas.getContext('2d') as CanvasRenderingContext2D;
         this.paths = new Array<Path2D>(this.lines.from.length);
```

Comment:
Oops, this is a huge memory waste, can't we simply use a function to determine whether a point belongs to a line? I'm sure we have it somewhere.

Reply:
Removed Path2D objects at all. Manipulating only on canvas context.